Add simple POST-command and changed upload path (##remote)

ogre2007 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (test plan)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (will do after rz-pipe pr)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

remote files can be uploaded now via /upload/<filename>, because root of the POST request path (/<filename>) shoud not be used for this
commands now can be transmitted via POST body, because its not good to transmit special characters via URI path (allows pipelinig, etc.)
should be enough to update rz-pipe in order to enable remote analysis

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

start rizin as an HTTP-service:
`rizin -q -e http.port=8080 -e http.root=$(pwd) -e http.upload=1 -e http.uproot=$(pwd) -cRh`
upload file as "temp":
`curl -si http://localhost:8080/upload/temp -F upload=@/bin/true`
analyze uploaded file:
`curl -si http://localhost:8080/cmd/ -d "o temp;pd  @ main;o-*"`
and see the result.
**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

idk

ogre2007 committed:
Add simple POST-command command and changed upload path (##remote)
remote files can be uploaded now via /upload/
commands now can be transmitted via POST body, because its not good to transmit special characters via URI path
should be enough to update rz-pipe in order to enable remote analysis

ogre2007 committed:
Merge branch 'dev' into http-post-cmd

Commit:
formatted with clang-format

XVilka commented:
Please add a test

XVilka commented on librz/core/rtr_http.c:
```diff
@@ -432,31 +438,51 @@ static int rz_core_rtr_http_run(RzCore *core, int launch, int browse, const char
-						snprintf(buf, sizeof(buf),
-							"<html><body><h2>uploaded %d byte(s). Thanks</h2>\n", retlen);
-						rz_socket_http_response(rs, 200, buf, 0, headers);
+			if (!strncmp(rs->path, "/upload/", 8)) {
```
While you are on it, I wonder if you are willing to also split this giant function into multiple. Something like router + handlers style.

ogre2007 committed:
Add test for http POST-cmd and upload feature
usage: python3 test/scripts/http_post_cmd_upload.py

XVilka commented:
You need also to run it as part of the rz-test, see e.g. `test/db/cmd/cmd_pipe`. Please also run `black` on this Python file.

ret2libc reviewed (changes requested):
Blocking this to discuss further:

why do we need this? Isn't it the same as moving the file with `nc` or with some other ways?

XVilka commented:
@ret2libc I think the idea is to allow the upload file to the remotely opened Rizin session.

ogre2007 commented:
> Blocking this to discuss further:
> 
> why do we need this? Isn't it the same as moving the file with `nc` or with some other ways?

The idea behind this is to allow uploading data (via rz-pipe with further changes, for example), and analyze it. However, this functions already were in r2/rizin, I've just refactored them a little now and added a feature that allows transmitting command not over GET-request path, but in POST-request body, which is more handy.

ret2libc commented:
> I've just refactored them a little now and added a feature that allows transmitting command not over GET-request path, but in POST-request body, which is more handy.

This seems better indeed, thanks for that!

> I think the idea is to allow the upload file to the remotely opened Rizin session.

@XVilka what's the difference with doing a separate `nc`(or something) to upload the binary and then analyzing it? You are assuming anyway that on the test machine there is a way to send a POST request to a remote server.

ogre2007 committed:
Merge branch 'dev' into http-post-cmd

ogre2007 committed:
Add test and split large http function

ogre2007 committed:
Pylint/black checkers fix

ogre2007 committed:
Fixed clang-format error on c file...

ogre2007 committed:
Merge branch 'dev' into http-post-cmd

ogre2007 committed:
Merge branch 'rizinorg:dev' into http-post-cmd

Commit:
Switched test from curl to requests

XVilka commented:
You need to install them first:
```
[XX] db/cmd/cmd_http_post http-post-cmd
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!python3 scripts/http_post_cmd_upload.py
' =
-- stdout
--- expected
+++ actual
@@ -1,2 +1,0 @@
-New and old cmd results equal
-Test succeeded

-- stderr
Traceback (most recent call last):
  File "/Users/runner/work/rizin/rizin/test/scripts/http_post_cmd_upload.py", line 14, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

```
python -m pip install requests
````

XVilka commented on librz/core/rtr_http.c:
```diff
@@ -3,6 +3,311 @@
+	if (rs->path[strlen(rs->path) - 1] == '/') {
+		path = (*index == '/') ? strdup(index) : rz_str_append(path, index);
+	} else {
+		//snprintf (path, sizeof (path), "%s/%s", root, rs->path);
```
Remove all unused code also.

ogre2007 committed:
Merge branch 'dev' into http-post-cmd

Commit:
Install requests in test

XVilka commented on test/db/cmd/cmd_http_post:
```diff
@@ -0,0 +1,11 @@
+NAME=http-post-cmd
+FILE==
+CMDS=<<EOF
+!python3 -m pip install requests
```
No, this is not the way to do it. Please update these files instead:
- `.builds/*`
- `.github/workflows/*`
- `.appveyor.yml`
- `.travis.yml`

ogre2007 committed:
Merge branch 'dev' into http-post-cmd

Commit:
Tried to add requests installing in github workflows

ret2libc commented on librz/core/rtr_http.c:
```diff
@@ -14,7 +311,7 @@ static int rz_core_rtr_http_run(RzCore *core, int launch, int browse, const char
 	int iport;
 	const char *host = rz_config_get(core->config, "http.bind");
-	const char *index = rz_config_get(core->config, "http.index");
+	//const char *index = rz_config_get(core->config, "http.index");
```
remove instead of commenting

ogre2007 committed:
Merge branch 'rizinorg:dev' into http-post-cmd

ogre2007 committed:
Removed comments

XVilka committed:
Merge branch 'dev' into http-post-cmd

XVilka commented:
Use files from `test/bins/` instead:
```

[XX] db/cmd/cmd_http_post http-post-cmd
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!python3 scripts/http_post_cmd_upload.py
' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-Test succeeded

-- stderr
Traceback (most recent call last):
  File "/Users/runner/work/rizin/rizin/test/scripts/http_post_cmd_upload.py", line 85, in <module>
    main()
  File "/Users/runner/work/rizin/rizin/test/scripts/http_post_cmd_upload.py", line 52, in main
    with open(TARGET, "rb") as file:
FileNotFoundError: [Errno 2] No such file or directory: '/bin/true'

[**]                               db/cmd/cmd_http_post    14372 OK       961 BR        1 XX       22 FX
```

XVilka commented:
@ogre2007 any plans to finish this?
